latch/functions/operators.py

```diff
@@ -0,0 +1,134 @@
+"""
+Mimics channel operators from Nextflow, using the correspondence Channel --> Python Dictionary
+"""
+
+import re
+from itertools import product
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+
+def _combine(item1: Any, item2: Any):
+    """
+    Combines two items for use in *_join functions. The rules followed are:
+
+    - If both items are lists, the lists are concatenated
+    - If one of the items is a list, the other is appended to that list
+    - Otherwise, the output is a new list containing both items
+
+    This is so that composition of joins works as expected. We also use list
+    addition so as to not modify the input items and instead return a new copy.
+    """
+
+    if isinstance(item1, List) and isinstance(item2, List):
+        return item1 + item2
+    elif isinstance(item1, List):
+        return item1 + [item2]
+    elif isinstance(item2, List):
+        return [item1] + item2
+    else:
+        return [item1, item2]
+
+
+def left_join(left: Dict[str, Any], right: Dict[str, Any]) -> Dict[str, Any]:
+    """A standard left join of two dictionaries, joining on their keys"""
+    output = {}
+    for key in left:
+        if key in right:
+            output[key] = _combine(left[key], right[key])
+        else:
+            output[key] = left[key]
+    return output
+
+
+def right_join(left: Dict[str, Any], right: Dict[str, Any]) -> Dict[str, Any]:
+    """A standard right join of two dictionaries, joining on their keys"""
+    output = {}
+    for key in right:
+        if key in left:
+            output[key] = _combine(left[key], right[key])
+        else:
+            output[key] = right[key]
+    return output
+
+
+def inner_join(left: Dict[str, Any], right: Dict[str, Any]) -> Dict[str, Any]:
+    """A standard inner join of two dictionaries, joining on their keys"""
+    output = {}
+    for key in left:
+        if key in right:
+            output[key] = _combine(left[key], right[key])
+    return output
+
+
+def outer_join(left: Dict[str, Any], right: Dict[str, Any]) -> Dict[str, Any]:
+    """A standard outer join of two dictionaries, joining on their keys"""
+    output = {}
+    for key in left:
+        if key in right:
+            output[key] = _combine(left[key], right[key])
+        else:
+            output[key] = left[key]
+    for key in right:
+        if key not in output:
+            output[key] = right[key]
+    return output
+
+
+def group_tuple(channel: List[Tuple], key_index: Optional[int] = None) -> List[Tuple]:
```

Comment:
took me a while to understand what was going on here, an example in the docstring would be helpful